// @sap-cloud-sdk/core/dist/odata/v2/request-builder/response-transformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformReturnValueForEdmTypeList = exports.transformReturnValueForEdmTypeListV2 = exports.transformReturnValueForEdmType = exports.transformReturnValueForEdmTypeV2 = exports.transformReturnValueForComplexTypeList = exports.transformReturnValueForComplexTypeListV2 = exports.transformReturnValueForComplexType = exports.transformReturnValueForComplexTypeV2 = exports.transformReturnValueForEntityList = exports.transformReturnValueForEntityListV2 = exports.transformReturnValueForEntity = exports.transformReturnValueForEntityV2 = exports.transformReturnValueForUndefined = exports.transformReturnValueForUndefinedV2 = void 0;
var entity_deserializer_1 = require("../entity-deserializer");
var response_data_accessor_1 = require("./response-data-accessor");
/* eslint-disable valid-jsdoc */
/**
 * @hidden
 */
function transformReturnValueForUndefinedV2(data, builderFn) {
    return builderFn(data);
}
exports.transformReturnValueForUndefinedV2 = transformReturnValueForUndefinedV2;
exports.transformReturnValueForUndefined = transformReturnValueForUndefinedV2;
/**
 * @hidden
 */
function transformReturnValueForEntityV2(data, entityConstructor) {
    return entity_deserializer_1.deserializeEntityV2(response_data_accessor_1.getSingleResult(data), entityConstructor).setOrInitializeRemoteState();
}
exports.transformReturnValueForEntityV2 = transformReturnValueForEntityV2;
exports.transformReturnValueForEntity = transformReturnValueForEntityV2;
/**
 * @hidden
 */
function transformReturnValueForEntityListV2(data, entityConstructor) {
    return response_data_accessor_1.getCollectionResult(data).map(function (entityJson) {
        return entity_deserializer_1.deserializeEntityV2(entityJson, entityConstructor).setOrInitializeRemoteState();
    });
}
exports.transformReturnValueForEntityListV2 = transformReturnValueForEntityListV2;
exports.transformReturnValueForEntityList = transformReturnValueForEntityListV2;
/**
 * @hidden
 */
function transformReturnValueForComplexTypeV2(data, builderFn) {
    return builderFn(response_data_accessor_1.getSingleResult(data));
}
exports.transformReturnValueForComplexTypeV2 = transformReturnValueForComplexTypeV2;
exports.transformReturnValueForComplexType = transformReturnValueForComplexTypeV2;
/**
 * @hidden
 */
function transformReturnValueForComplexTypeListV2(data, builderFn) {
    return response_data_accessor_1.getCollectionResult(data).map(function (json) { return builderFn(json); });
}
exports.transformReturnValueForComplexTypeListV2 = transformReturnValueForComplexTypeListV2;
exports.transformReturnValueForComplexTypeList = transformReturnValueForComplexTypeListV2;
/**
 * @hidden
 */
function transformReturnValueForEdmTypeV2(data, builderFn) {
    return builderFn(response_data_accessor_1.getSingleResult(data));
}
exports.transformReturnValueForEdmTypeV2 = transformReturnValueForEdmTypeV2;
exports.transformReturnValueForEdmType = transformReturnValueForEdmTypeV2;
/**
 * @hidden
 */
function transformReturnValueForEdmTypeListV2(data, builderFn) {
    return response_data_accessor_1.getCollectionResult(data).map(builderFn);
}
exports.transformReturnValueForEdmTypeListV2 = transformReturnValueForEdmTypeListV2;
exports.transformReturnValueForEdmTypeList = transformReturnValueForEdmTypeListV2;
