// @sap-cloud-sdk/core/dist/scp-cf/destination/destination-from-service.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDestinationFromDestinationService = void 0;
var util_1 = require("@sap-cloud-sdk/util");
var util_2 = require("../../util");
var cache_1 = require("../cache");
var token_accessor_1 = require("../token-accessor");
var connectivity_service_1 = require("../connectivity-service");
var environment_accessor_1 = require("../environment-accessor");
var destination_selection_strategies_1 = require("./destination-selection-strategies");
var destination_service_1 = require("./destination-service");
var destination_cache_1 = require("./destination-cache");
var logger = util_1.createLogger({
    package: 'core',
    messageContext: 'destination-accessor-service'
});
var emptyDestinationByType = {
    instance: [],
    subaccount: []
};
/**
 * Retrieves a destination with the given name from the Cloud Foundry destination service.
 * Returns null if no destination can be found.
 * Requires the following service bindings: destination, XSUAA
 * By default, selects subscriber over provider and instance over subaccount destinations.
 *
 * If the destinations are read from the environment, the jwt will be ignored.
 *
 * @param name - The name of the destination to be retrieved.
 * @param options - Configuration for how to retrieve destinations from the destination service.
 * @returns A promise returning the requested destination on success.
 */
function getDestinationFromDestinationService(name, options) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, DestinationFromServiceRetriever.getDestinationFromDestinationService(name, options)];
        });
    });
}
exports.getDestinationFromDestinationService = getDestinationFromDestinationService;
var DestinationFromServiceRetriever = /** @class */ (function () {
    function DestinationFromServiceRetriever(name, options, decodedUserJwt, providerToken) {
        this.name = name;
        this.decodedUserJwt = decodedUserJwt;
        this.providerToken = providerToken;
        this.options = __assign({}, options);
        this.decodedProviderToken = util_2.decodeJwt(providerToken);
        this.options.isolationStrategy =
            options.isolationStrategy || cache_1.IsolationStrategy.Tenant;
        this.options.selectionStrategy =
            options.selectionStrategy || destination_selection_strategies_1.subscriberFirst;
        this.options.useCache = options.useCache || false;
    }
    DestinationFromServiceRetriever.getDestinationFromDestinationService = function (name, options) {
        return __awaiter(this, void 0, void 0, function () {
            var decodedUserJwt, providerToken, da, destinationResult, destination, withProxySetting;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, DestinationFromServiceRetriever.getDecodedUserJwt(options)];
                    case 1:
                        decodedUserJwt = _a.sent();
                        return [4 /*yield*/, DestinationFromServiceRetriever.getProviderToken(options)];
                    case 2:
                        providerToken = _a.sent();
                        da = new DestinationFromServiceRetriever(name, options, decodedUserJwt, providerToken);
                        return [4 /*yield*/, da.searchDestinationWithSelectionStrategyAndCache()];
                    case 3:
                        destinationResult = _a.sent();
                        if (!destinationResult) {
                            return [2 /*return*/, null];
                        }
                        if (destinationResult.fromCache) {
                            return [2 /*return*/, destinationResult.destination];
                        }
                        destination = destinationResult.destination;
                        if (!(destination.authentication === 'OAuth2SAMLBearerAssertion')) return [3 /*break*/, 5];
                        return [4 /*yield*/, da.addOAuthSamlAuth(destination)];
                    case 4:
                        destination = _a.sent();
                        _a.label = 5;
                    case 5:
                        if (!(destination.authentication === 'ClientCertificateAuthentication')) return [3 /*break*/, 7];
                        return [4 /*yield*/, da.addClientCertAuth()];
                    case 6:
                        destination = _a.sent();
                        _a.label = 7;
                    case 7: return [4 /*yield*/, da.addProxyConfiguration(destination)];
                    case 8:
                        withProxySetting = _a.sent();
                        da.updateDestinationCache(withProxySetting, destinationResult.origin);
                        return [2 /*return*/, withProxySetting];
                }
            });
        });
    };
    DestinationFromServiceRetriever.getDecodedUserJwt = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, options.userJwt
                        ? util_2.verifyJwt(options.userJwt, options)
                        : options.iss
                            ? { iss: options.iss }
                            : undefined];
            });
        });
    };
    DestinationFromServiceRetriever.getProviderToken = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var userJwt, optionsWithoutJwt;
            return __generator(this, function (_a) {
                userJwt = options.userJwt, optionsWithoutJwt = __rest(options, ["userJwt"]);
                return [2 /*return*/, token_accessor_1.serviceToken('destination', optionsWithoutJwt)];
            });
        });
    };
    DestinationFromServiceRetriever.prototype.searchDestinationWithSelectionStrategyAndCache = function () {
        return __awaiter(this, void 0, void 0, function () {
            var destinationSearchResult;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.isSubscriberNeeded()) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.searchSubscriberAccountForDestination()];
                    case 1:
                        destinationSearchResult = _a.sent();
                        _a.label = 2;
                    case 2:
                        if (!this.isProviderNeeded(destinationSearchResult)) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.searchProviderAccountForDestination()];
                    case 3:
                        destinationSearchResult = _a.sent();
                        _a.label = 4;
                    case 4:
                        if (destinationSearchResult && !destinationSearchResult.fromCache) {
                            logger.info('Successfully retrieved destination from destination service.');
                        }
                        if (destinationSearchResult && destinationSearchResult.fromCache) {
                            logger.info("Successfully retrieved destination from destination service cache for " + destinationSearchResult.origin + " destinations.");
                        }
                        if (!destinationSearchResult) {
                            logger.info('Could not retrieve destination from destination service.');
                        }
                        return [2 /*return*/, destinationSearchResult];
                }
            });
        });
    };
    DestinationFromServiceRetriever.prototype.getInstanceAndSubaccountDestinations = function (accessToken) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, instance, subaccount;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, Promise.all([
                            destination_service_1.fetchInstanceDestinations(this.destinationServiceCredentials.uri, accessToken, this.options),
                            destination_service_1.fetchSubaccountDestinations(this.destinationServiceCredentials.uri, accessToken, this.options)
                        ])];
                    case 1:
                        _a = _b.sent(), instance = _a[0], subaccount = _a[1];
                        return [2 /*return*/, {
                                instance: instance,
                                subaccount: subaccount
                            }];
                }
            });
        });
    };
    Object.defineProperty(DestinationFromServiceRetriever.prototype, "destinationServiceCredentials", {
        get: function () {
            var credentials = environment_accessor_1.getDestinationServiceCredentialsList();
            if (!credentials || credentials.length === 0) {
                throw Error('No binding to a Destination service instance found. Please bind a destination service instance to your application!');
            }
            return credentials[0];
        },
        enumerable: false,
        configurable: true
    });
    DestinationFromServiceRetriever.prototype.addOAuthSamlAuth = function (destination) {
        return __awaiter(this, void 0, void 0, function () {
            var destinationService, accessToken;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        destinationService = environment_accessor_1.getService('destination');
                        if (!destinationService) {
                            throw Error("Failed to fetch destination \"" + this.name + "\"! No binding to a destination service found.");
                        }
                        if (destination.systemUser) {
                            logger.debug("System user found on destination. The provider token: " + this.providerToken + " is used for destination fetching");
                            return [2 /*return*/, destination_service_1.fetchDestination(destinationService.credentials.uri, this.providerToken, this.name, this.options)];
                        }
                        if (!this.options.userJwt) {
                            throw Error('No user token (JWT) has been provided! This is strictly necessary for principal propagation.');
                        }
                        return [4 /*yield*/, token_accessor_1.userApprovedServiceToken(this.options.userJwt, destinationService, this.options)];
                    case 1:
                        accessToken = _a.sent();
                        return [2 /*return*/, destination_service_1.fetchDestination(destinationService.credentials.uri, accessToken, this.name, this.options)];
                }
            });
        });
    };
    DestinationFromServiceRetriever.prototype.addClientCertAuth = function () {
        return __awaiter(this, void 0, void 0, function () {
            var accessToken;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, token_accessor_1.serviceToken('destination', __assign({ userJwt: this.decodedUserJwt }, this.options))];
                    case 1:
                        accessToken = _a.sent();
                        return [2 /*return*/, destination_service_1.fetchDestination(this.destinationServiceCredentials.uri, accessToken, this.name, this.options)];
                }
            });
        });
    };
    DestinationFromServiceRetriever.prototype.addProxyConfiguration = function (destination) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (util_2.proxyStrategy(destination)) {
                    case util_2.ProxyStrategy.ON_PREMISE_PROXY:
                        return [2 /*return*/, connectivity_service_1.addProxyConfigurationOnPrem(destination, this.options.userJwt)];
                    case util_2.ProxyStrategy.INTERNET_PROXY:
                        return [2 /*return*/, util_2.addProxyConfigurationInternet(destination)];
                    case util_2.ProxyStrategy.NO_PROXY:
                        return [2 /*return*/, destination];
                    default:
                        throw new Error('Illegal argument: No valid proxy configuration found in the destination input to be aded.');
                }
                return [2 /*return*/];
            });
        });
    };
    DestinationFromServiceRetriever.prototype.updateDestinationCache = function (destination, destinationOrigin) {
        if (!this.options.useCache) {
            return destination;
        }
        destination_cache_1.destinationCache.cacheRetrievedDestination(destinationOrigin === 'subscriber'
            ? this.decodedUserJwt
            : this.decodedProviderToken, destination, this.options.isolationStrategy);
    };
    DestinationFromServiceRetriever.prototype.getProviderDestinationService = function () {
        return __awaiter(this, void 0, void 0, function () {
            var provider, destination;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getInstanceAndSubaccountDestinations(this.providerToken)];
                    case 1:
                        provider = _a.sent();
                        return [4 /*yield*/, this.options.selectionStrategy({
                                subscriber: emptyDestinationByType,
                                provider: provider
                            }, this.name)];
                    case 2:
                        destination = _a.sent();
                        if (destination) {
                            return [2 /*return*/, {
                                    destination: destination,
                                    fromCache: false,
                                    origin: 'provider'
                                }];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    DestinationFromServiceRetriever.prototype.getProviderDestinationCache = function () {
        var destination = destination_cache_1.destinationCache.retrieveDestinationFromCache(this.decodedProviderToken, this.name, this.options.isolationStrategy);
        if (destination) {
            return { destination: destination, fromCache: true, origin: 'subscriber' };
        }
    };
    DestinationFromServiceRetriever.prototype.getSubscriberDestinationService = function () {
        return __awaiter(this, void 0, void 0, function () {
            var accessToken, subscriber, destination;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, token_accessor_1.serviceToken('destination', __assign({ userJwt: this.decodedUserJwt }, this.options))];
                    case 1:
                        accessToken = _a.sent();
                        return [4 /*yield*/, this.getInstanceAndSubaccountDestinations(accessToken)];
                    case 2:
                        subscriber = _a.sent();
                        destination = this.options.selectionStrategy({
                            subscriber: subscriber,
                            provider: emptyDestinationByType
                        }, this.name);
                        if (destination) {
                            return [2 /*return*/, { destination: destination, fromCache: false, origin: 'subscriber' }];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    DestinationFromServiceRetriever.prototype.getSubscriberDestinationCache = function () {
        var destination = destination_cache_1.destinationCache.retrieveDestinationFromCache(this.decodedUserJwt, this.name, this.options.isolationStrategy);
        if (destination) {
            return { destination: destination, fromCache: true, origin: 'subscriber' };
        }
    };
    DestinationFromServiceRetriever.prototype.isProviderAndSubscriberSameTenant = function () {
        return (this.decodedUserJwt &&
            util_2.isIdenticalTenant(this.decodedUserJwt, this.decodedProviderToken));
    };
    DestinationFromServiceRetriever.prototype.isProviderNeeded = function (resultFromSubscriber) {
        if (this.options.selectionStrategy === destination_selection_strategies_1.alwaysSubscriber) {
            return false;
        }
        if (this.options.selectionStrategy === destination_selection_strategies_1.subscriberFirst &&
            resultFromSubscriber) {
            return false;
        }
        return true;
    };
    DestinationFromServiceRetriever.prototype.isSubscriberNeeded = function () {
        if (!this.decodedUserJwt) {
            return false;
        }
        if (this.options.selectionStrategy === destination_selection_strategies_1.alwaysProvider) {
            return false;
        }
        if (this.isProviderAndSubscriberSameTenant()) {
            return false;
        }
        return true;
    };
    DestinationFromServiceRetriever.prototype.searchProviderAccountForDestination = function () {
        return __awaiter(this, void 0, void 0, function () {
            var destination;
            return __generator(this, function (_a) {
                if (this.options.useCache) {
                    destination = this.getProviderDestinationCache();
                }
                return [2 /*return*/, destination || this.getProviderDestinationService()];
            });
        });
    };
    DestinationFromServiceRetriever.prototype.searchSubscriberAccountForDestination = function () {
        return __awaiter(this, void 0, void 0, function () {
            var destination;
            return __generator(this, function (_a) {
                if (this.options.useCache) {
                    destination = this.getSubscriberDestinationCache();
                }
                return [2 /*return*/, destination || this.getSubscriberDestinationService()];
            });
        });
    };
    return DestinationFromServiceRetriever;
}());
