// @sap-cloud-sdk/core/dist/odata/v2/uri-conversion/odata-uri.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oDataUri = exports.oDataUriV2 = void 0;
var common_1 = require("../../common");
var get_expand_1 = require("./get-expand");
var get_select_1 = require("./get-select");
var uri_value_converter_1 = require("./uri-value-converter");
var getFilter = common_1.createGetFilter(uri_value_converter_1.uriConverterV2).getFilter;
var getResourcePathForKeys = common_1.createGetResourcePathForKeys(uri_value_converter_1.uriConverterV2).getResourcePathForKeys;
var convertToUriFormat = uri_value_converter_1.uriConverterV2.convertToUriFormat;
/**
 * Instance of the [[ODataUri]] conversion interface for OData v2.
 */
exports.oDataUriV2 = {
    getExpand: get_expand_1.getExpandV2,
    getFilter: getFilter,
    getEntityKeys: common_1.getEntityKeys,
    getOrderBy: common_1.getOrderBy,
    getResourcePathForKeys: getResourcePathForKeys,
    getSelect: get_select_1.getSelectV2,
    convertToUriFormat: convertToUriFormat
};
exports.oDataUri = exports.oDataUriV2;
