// @sap/cds-runtime/lib/messaging/file-based.js
const { resolve } = require('path')
const { emit, init, watch } = require('./file-based-utils/client')
const cds = global.cds || require('@sap/cds/lib')

const MessagingService = require('./service')
const _error = console.error

const DEBUG = cds.debug('messaging')

class FileBasedMessaging extends MessagingService {
  async init () {
    const file = this.options.file || (this.options.file = '~/.cds-msg-box')
    this.file = resolve(file.replace(/^~/, () => require('os').userInfo().homedir))
    this.lock = `${this.file}.lock`
    this.status = { active: true, lastCtimeMs: 0 }
    this.subscriptions = new Set()
    cds.once('listening', () => {
      this.ready = true
      this.listen()
    })

    await init(this.file, this.lock)
    super.on('*', (req, next) => {
      if (req.inbound) return next()
      DEBUG && DEBUG('Emit', { topic: req.event, file: this.file })
      if (this.options.outbox) req.on('succeeded', () => emit(req, this.file, this.lock))
      else return emit(req, this.file, this.lock)
    })
    return super.init()
  }

  // inbound -> listen to channel (once)
  async subscribe (phase, event, path, handler) {
    const topic = event
    this.subscriptions.add(topic)
    if (this.subscriptions.size === 1) this.listen()
    return super.subscribe(phase, event, path, handler)
  }

  listen () {
    if (!this.listening && this.ready && this.subscriptions.size > 0) {
      this.listening = true
      // console.log('[cds] - Watch', { file: this.file }) // no std trace output by non-CLI components please
      watch(this.file, this.lock, this.subscriptions, this.status, async msg => {
        try {
          await this.emit({ ...msg, inbound: true })
        } catch (e) {
          _error(e)
        }
      })
    }
  }

  disconnect () {
    this.status.active = false
  }
}

module.exports = FileBasedMessaging
