// @sap/cds-runtime/lib/common/utils/flavor.js
// REVISIT: rm once cds.env.effective set by umbrella

const cds = global.cds || require('@sap/cds/lib')

const w4 = {
  version: 'v4',
  containment: true,
  structs: true,
  refs: false,
  format: 'structured',
  _foreignKeys: true,
  __proxies: false,
  odataFormat: 'structured',
  odataContainment: true,
  odataProxies: false
}
const x4 = { ...w4, refs: true, proxies: true, _foreignKeys: false, __proxies: true, odataProxies: true }
const flavors = { w4, x4 }

module.exports = () => {
  if (!cds.env.effective) cds.env.effective = {}
  if (!cds.env.effective.odata) cds.env.effective.odata = { ...cds.env.odata }

  const flavor = cds.env.odata.flavor
  if (flavor in flavors) {
    Object.assign(cds.env.odata, flavors[flavor]) // for compat _foreignKeys and __proxies
    Object.assign(cds.env.effective.odata, cds.env.odata)

    // okra
    cds.env.odata.skipValidation = true
  }
}

// REVISIT: this should be removable
const { odata } = cds.env
if (odata.flavors && odata.flavors.x4.structs) {
  module.exports = () => {
    if (odata.flavor in odata.flavors) odata.skipValidation = true
  }
}
