// @sap-cloud-sdk/core/dist/odata/v2/entity-serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeEntityNonCustomFields = exports.serializeComplexType = exports.serializeEntity = exports.serializeEntityNonCustomFieldsV2 = exports.serializeComplexTypeV2 = exports.serializeEntityV2 = exports.entitySerializerV2 = void 0;
var entity_serializer_1 = require("../common/entity-serializer");
var payload_value_converter_1 = require("./payload-value-converter");
/**
 * Entity serializer instance for v2 entities.
 * See [[EntitySerializer]] for the provided methods.
 */
exports.entitySerializerV2 = entity_serializer_1.entitySerializer(payload_value_converter_1.tsToEdmV2);
exports.serializeEntityV2 = exports.entitySerializerV2.serializeEntity;
exports.serializeEntity = exports.serializeEntityV2;
exports.serializeComplexTypeV2 = exports.entitySerializerV2.serializeComplexType;
exports.serializeComplexType = exports.serializeComplexTypeV2;
exports.serializeEntityNonCustomFieldsV2 = exports.entitySerializerV2.serializeEntityNonCustomFields;
exports.serializeEntityNonCustomFields = exports.serializeEntityNonCustomFieldsV2;
