// @sap-cloud-sdk/core/dist/odata/v4/request-builder/response-transformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformReturnValueForEdmTypeListV4 = exports.transformReturnValueForEdmTypeV4 = exports.transformReturnValueForComplexTypeListV4 = exports.transformReturnValueForComplexTypeV4 = exports.transformReturnValueForEntityListV4 = exports.transformReturnValueForEntityV4 = exports.transformReturnValueForUndefinedV4 = void 0;
var entity_deserializer_1 = require("../entity-deserializer");
var response_data_accessor_1 = require("./response-data-accessor");
/* eslint-disable valid-jsdoc */
/**
 * @hidden
 */
function transformReturnValueForUndefinedV4(data, builderFn) {
    return builderFn(data);
}
exports.transformReturnValueForUndefinedV4 = transformReturnValueForUndefinedV4;
/**
 * @hidden
 */
function transformReturnValueForEntityV4(data, entityConstructor) {
    return entity_deserializer_1.deserializeEntityV4(response_data_accessor_1.getSingleResult(data), entityConstructor).setOrInitializeRemoteState();
}
exports.transformReturnValueForEntityV4 = transformReturnValueForEntityV4;
/**
 * @hidden
 */
function transformReturnValueForEntityListV4(data, entityConstructor) {
    return response_data_accessor_1.getCollectionResult(data).map(function (entityJson) {
        return entity_deserializer_1.deserializeEntityV4(entityJson, entityConstructor).setOrInitializeRemoteState();
    });
}
exports.transformReturnValueForEntityListV4 = transformReturnValueForEntityListV4;
/**
 * @hidden
 */
function transformReturnValueForComplexTypeV4(data, builderFn) {
    return builderFn(response_data_accessor_1.getSingleResult(data));
}
exports.transformReturnValueForComplexTypeV4 = transformReturnValueForComplexTypeV4;
/**
 * @hidden
 */
function transformReturnValueForComplexTypeListV4(data, builderFn) {
    return response_data_accessor_1.getCollectionResult(data).map(function (json) { return builderFn(json); });
}
exports.transformReturnValueForComplexTypeListV4 = transformReturnValueForComplexTypeListV4;
/**
 * @hidden
 */
function transformReturnValueForEdmTypeV4(data, builderFn) {
    return builderFn(response_data_accessor_1.getSingleResult(data));
}
exports.transformReturnValueForEdmTypeV4 = transformReturnValueForEdmTypeV4;
/**
 * @hidden
 */
function transformReturnValueForEdmTypeListV4(data, builderFn) {
    return response_data_accessor_1.getCollectionResult(data).map(builderFn);
}
exports.transformReturnValueForEdmTypeListV4 = transformReturnValueForEdmTypeListV4;
